src/converter/one/heatmap/Heatmap.java

```diff
@@ -0,0 +1,465 @@
+/*
+ * Copyright The async-profiler authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package one.heatmap;
+
+import java.io.PrintStream;
+import one.convert.Arguments;
+import one.jfr.ClassRef;
+import one.jfr.Dictionary;
+import one.jfr.MethodRef;
+import one.jfr.DictionaryInt;
+import one.jfr.Index;
+import one.convert.ResourceProcessor;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Arrays;
+import java.util.Comparator;
+
+public class Heatmap {
+
+    private final Arguments arguments;
+
+    private State state;
+
+    private long startMs;
+
+    public Heatmap(Arguments arguments) {
+        this.arguments = arguments;
+
+        // TODO it is probably should be an argument
+        // but there is a good chance that changing it will have some side effects
+        this.state = new State(20);
+    }
+
+    public void assignConstantPool(
+            Dictionary<MethodRef> methodRefs,
+            Dictionary<ClassRef> classRefs,
+            Dictionary<byte[]> symbols
+    ) {
+        state.methodsCache.assignConstantPool(methodRefs, classRefs, symbols);
+    }
+
+    public void nextFile() {
+        state.stackTracesCache.clear();
+        state.methodsCache.clear();
+    }
+
+    public void addEvent(int stackTraceId, int extra, byte type, long timeMs) {
+        state.addEvent(stackTraceId, extra, type, timeMs);
+    }
+
+    public void addStack(long id, long[] methods, int[] locations, byte[] types, int size) {
+        state.addStack(id, methods, locations, types, size);
+    }
+
+    public void finish(long startMs) {
+        this.startMs = startMs;
+        assignConstantPool(null, null, null);
+        nextFile();
+    }
+
+    private EvaluationContext evaluate() {
+        State state = this.state;
+        this.state = null;
+        return new EvaluationContext(
+                state.sampleList.samples(),
+                state.methodsCache.methodsIndex(),
+                state.stackTracesRemap.orderedTraces(),
+                state.methodsCache.orderedSymbolTable()
+        );
+    }
+
+    private void compressMethods(HtmlOut out, Method[] methods) throws IOException {
+        for (Method method : methods) {
+            out.write18(method.className);
+            out.write18(method.methodName);
```

Comment:
So, the conversion will fail if there are more than 2^18 methods?
Also, will it be more compact to use varint encoding for classess, methods and locations?

Reply:
It can be more compact, yes, but using var length encoding may lead to extra preprocessing step on the html side. I'll check if it is the case.
But it should be more then 2^18 unique method names (not including classes). Anyway, we don't need to be compact here, but it would be nice to be safe)

Reply:
refactored